sources/mma/mma.f90

```diff
@@ -0,0 +1,701 @@
+        allocate(this%pij(m,n))
+        if (allocated(this%qij)) deallocate(this%qij) 
+        allocate(this%qij(m,n))
+        if (allocated(this%bi)) deallocate(this%bi) 
+        allocate(this%bi(m))
+        !---nesessary for KKT check after updating df0dx, fval, dfdx --------
+        if (allocated(this%y)) deallocate(this%y) 
+        allocate(this%y(m))
+        if (allocated(this%lambda)) deallocate(this%lambda) 
+        allocate(this%lambda(m))
+        if (allocated(this%s)) deallocate(this%s) 
+        allocate(this%s(m))
+        if (allocated(this%mu)) deallocate(this%mu) 
+        allocate(this%mu(m))
+        if (allocated(this%xsi)) deallocate(this%xsi) 
+        allocate(this%xsi(n))
+        if (allocated(this%eta)) deallocate(this%eta) 
+        allocate(this%eta(n))
+
+
+        this%epsimin =  1.0e-10 !0.0000001
+        
+        this%max_iter = 100
+
+        !!!!!!! For TEST_CASE_1
+        this%a0=1
+        this%a=0
+        this%c=1000
+        this%d=0
+        !setting the bounds for the design variable based on the problem
+        this%xmax(:) = 10
+        this%xmin(:) = 0 !1e-8 !0.00000001
+
+        !!!!!!! For TEST_CASE_2
+        ! this%a0=1.0
+        ! this%a=0
+        ! this%c=1000
+        ! this%d=1
+        ! !setting the bounds for the design variable based on the problem
+        ! this%xmax(:) = 5
+        ! this%xmin(:) = 0
+
+
+
+
+        this%low(:) = minval(x)
+        this%upp(:) = maxval(x)
+
+        !following parameters are set based on eq.3.8:--------
+        this%asyinit = 0.5 !
+        this%asyincr = 1.2 ! 1.1
+        this%asydecr = 0.7; !0.65
+
+        !setting KKT norms to a large number for the initial design
+        this%residumax = 10**5
+        this%residunorm = 10**5
+        
+        !the object is correctly initialized
+        this%is_initialized = .true.
+    end subroutine mma_init 
+
+    subroutine mma_update(this, iter, x, df0dx, fval, dfdx)
+        ! ----------------------------------------------------- !
+        ! Update the design variable of the mma object (this%x) !
+        ! by solving the convex approximation of the problem.   !
+        !                                                       !
+        ! This subroutine is called in each iteration of the    !
+        ! optimization loop                                     !
+        ! ----------------------------------------------------- !
+        class(mma_t), intent(inout) :: this
+        integer, intent(in) :: iter
+        real(kind=rp), dimension(this%n), intent(in) :: x
+        real(kind=rp), dimension(this%n), intent(in) :: df0dx(:)
+        real(kind=rp), dimension(this%m), intent(in) :: fval(:)
+        real(kind=rp), dimension(this%m, this%n), intent(in) :: dfdx(:,:)
+
+        if (.not.this%is_initialized) print*,"The MMA object is not &
+            initialized. Please call MMA_obj.init() first and then & 
+            call MMA_obj.update()."
```

Comment:
Please dont use a single line if statement, if it does not fit on a single line.
```suggestion
        if (.not.this%is_initialized) then
            print*,"The MMA object is not &
                initialized. Please call MMA_obj.init() first and then & 
                call MMA_obj.update()."
        end if
```